dataprep/clean/clean_duplication_utils.py

```diff
@@ -0,0 +1,323 @@
+        # create a column "vals" containing tuples of the form (item, count)
+        no_dups.loc[:, "vals"] = no_dups[col].map(lambda val: (val, counts.loc[val]))
+        # create a column "key" containing keys created by the given key collision method
+        no_dups.loc[:, "key"] = no_dups[col].map(key_func)
+        # put items with the same key into the same list
+        clusters = no_dups.groupby("key")["vals"].agg(list)
+        clusters = clusters.loc[clusters.map(len) > 1]
+        # values with greater counts appear first in the cluster
+        clusters = clusters.map(lambda x: sorted(x, key=itemgetter(1), reverse=True))
+        # sort by the size of each cluster, so that larger clusters appear first
+        # can be replaced by clusters.sort_values(key=lambda x: x.map(len), ascending=False) once
+        # required pandas version is updated
+        clusters_df = pd.DataFrame({"cluster": clusters, "len": clusters.map(len)})
+        clusters_df = clusters_df.sort_values(by="len", ascending=False)
+        self.clusters = clusters_df.loc[:, "cluster"]
+
+    def _nearest_neighbours_cluster(self) -> None:
+        """
+        Performs nearest neighbour clustering.
+        Blocking is used to speed up the process, blocks are obtained where strings in the same
+        block share a substring of a given blocking size. Only strings within the same block are
+        compared using the levenshtein distance function.
+
+        Method from OpenRefine: https://github.com/OpenRefine/OpenRefine/wiki/Clustering-In-Depth
+        and simile-vicino: https://code.google.com/archive/p/simile-vicino/
+        """
+        col = self._col
+        blocks: DefaultDict[str, Set[str]] = collections.defaultdict(set)
+        counts = self._df.groupby(col).size()
+        no_dups = self._df.drop_duplicates(subset=[col]).reset_index()
+
+        # put strings in blocks
+        no_dups[col].apply(self._populate_blocks, args=(blocks, self._block_size))
+        # compare strings in the same block and create clusters
+        self.clusters = self._get_nearest_neighbour_clusters(blocks, counts, self._radius)
+
+    @staticmethod
+    def _populate_blocks(val: str, blocks: DefaultDict[str, Set[str]], block_size: int) -> None:
+        """
+        Create n gram tokens of the given string and place the string into the block
+        for each n gram.
+        """
+        tokens = _ngram_tokens(val, block_size)
+        for token in tokens:
+            blocks[token].add(val)
+
+    @staticmethod
+    def _get_nearest_neighbour_clusters(
+        blocks: DefaultDict[str, Set[str]], counts: pd.Series, radius: int
+    ) -> pd.Series:
+        """
+        Compare every pair of strings in each block and add to cluster if
+        their distance is less than the given radius.
+        """
+        cluster_map = collections.defaultdict(set)
+        for block in blocks.values():
+            for str1 in block:
```

Comment:
If you want to iterate over all unique pairs of strings, have you considered `combinations` from `itertools`?

``` python
for str1, str2 in combinations(block, 2):
    ...
```
please ignore me if this is not what you want here.

Reply:
This part works by picking a center for each cluster and finding all the values within `radius` of the center. `str1` is the current center and every string has to have a turn being the center. Since the order matters I don't think `combinations` would work for my current approach. I could use `permutations(block, 2)` if you like it better than the double for loop?

Reply:
Got it. I think you can do whatever you like, but in my opinion `permutations` would make the code a bit more succinct/elegant compared to a double for loop.